### DISCUSSION TITLE
AI generate commit title and message

### DISCUSSION THREAD

**Discussion by pranav-dubey-cl** [MichaelCurrin/auto-commit-msg#109] [General]
Does this repo also supports AI generated commit message and title? I am thinking of a way with which a model can go through past commits tuning itself for a repo based on commit message and the changes in file to provide a better and relevant commit message with follows the format and flavors of the repo.

**Comment by MichaelCurrin** [comment]
See this note in my docs on AI-based extensions. Maybe find one that has the functionality you are looking for or add an issue there.

https://github.com/MichaelCurrin/auto-commit-msg/blob/master/docs/other/ai-tools.md

**Comment by MichaelCurrin** [comment]
I also created a CLI tool quickly in Python as a POC of about 100 lines. It runs against a local LLM using git diff and generates a commit message. I could share that as a tool, as least for the CLI and maybe rewrite as a new VS Code extension.

## Sample

Here is the prompt I used

### System prompt

```
    You are a helpful assistant that generates concise and informative Git commit messages based on the
    provided diff, describing changes to all the files in the diff and summarising the changes at a high-level if necessary.

    Provide a SINGLE commit message of ONE LINE, of length 50 to 72 characters, with optional description below,
    and return NOTHING else.

    Here is the conventional commit specficiation:

    - `feat` - feature
    - `fix`
    - `style`
    - `refactor`
    - `build` - a change in the build system.
    - `perf` - performance
    - `ci` - changes to the CI.
    - `docs` - changes to documentation files or docs in code.
    - `test` - relating to running tests.
    - `chore` - changes to configs, file renames or moves, changes to dependencies.
    - `revert`

    Create commits based on that.
    e.g. feat: Add foo to bar.
```

### User prompt

```
Generate a single one-line conventional commit message and nothing else, using the following Git diff:
```

then append git diff output.

---

With that, you could customize that to include your style of commits that are manually entered or read from the history.

(If your history is really long, you could also use a cleaning step of AI to extract the predominant style of messages and then feed the smaller result into the commit message generator.
